### ISSUE TITLE
Observation from a single example: original vs. debiased cross-image similarity maps look similar

### ISSUE THREAD

**Issue by Ginven-W** [visinf/INSID3#13]
First of all, thank you for this wonderful work and for making the code publicly available. I have a question regarding the effect of positional encoding in cross-image similarity.

Here is an example using two cats from the assets to examine this effect. To better highlight positional differences, I applied cropping to move the cat to different locations (top-left in the reference image and bottom-right in the target image), making the effect easier to observe.

<img width="2291" height="2823" alt="Image" src="https://github.com/user-attachments/assets/0918a1d1-68bd-455b-bdb1-4005ddf25907" />

In the third row, I visualize the similarity maps computed using original features and debiased features, respectively. Although it is possible to observe that the original features still produce some activation on the top-left, the effect is not very pronounced and does not match the level shown in the paper’s example (see the figure below).
<img width="383" height="122" alt="Image" src="https://github.com/user-attachments/assets/9fe47b8c-1137-4632-87c1-b183a185e5cc" />

My processing code is:

```python
s = ref_images.shape[0]
imgs = torch.cat([ref_images, tgt_image], dim=0).unsqueeze(0)
fmaps = model._extract_features(imgs)
fmaps_norm = F.normalize(fmaps, p=2, dim=2)
fmaps_deb = model._debias_features(fmaps_norm)

feat_refs_orig = fmaps_norm[:, :s]
feat_tgt_orig = fmaps_norm[:, s] 
feat_refs_deb = fmaps_deb[:, :s]
feat_tgt_deb = fmaps_deb[:, s]   

# refer prototype
ref_masks_4d = ref_masks.unsqueeze(1)  # (S,1,H,W)
proto_orig_list = []
proto_deb_list = []
for i in range(s):
    mask_i = downsample_mask(ref_masks_4d[i:i + 1], h, w)  # (1,h,w)
    fg_orig = feat_refs_orig[0, i, :, mask_i]
    fg_deb = feat_refs_deb[0, i, :, mask_i]
    if fg_orig.shape[1] > 0:
        proto_orig_list.append(fg_orig.mean(dim=1))
        proto_deb_list.append(fg_deb.mean(dim=1))
ref_proto_orig = F.normalize(torch.stack(proto_orig_list).mean(dim=0), p=2, dim=0)  # (C,)
ref_proto_deb = F.normalize(torch.stack(proto_deb_list).mean(dim=0), p=2, dim=0)    # (C,)

# compute cross similarity
sim_orig = torch.einsum("chw,c->hw", feat_tgt_orig[0], ref_proto_orig)
sim_deb = torch.einsum("chw,c->hw", feat_tgt_deb[0], ref_proto_deb)
sim_orig = _minmax01(sim_orig).unsqueeze(0).unsqueeze(0)
sim_deb = _minmax01(sim_deb).unsqueeze(0).unsqueeze(0)
sim_orig_up = F.interpolate(sim_orig, size=(h0, w0), mode="nearest").squeeze().cpu().numpy()
sim_deb_up = F.interpolate(sim_deb, size=(h0, w0), mode="nearest").squeeze().cpu().numpy()
```
I wonder if there is any error in my implementation or reproduction.

Thank you for your guidance!

**Comment by ClaudiaCuttano** [comment]
Hi @Ginven-W, thank you for your interest!

As discussed in the paper, we found this positional component to be a consistent property of DINOv3 dense features across the model scales we tested, including small, base, and large models. However, the visual strength of the effect is not uniform: it depends on the model size, the image pair, the queried region, and how much semantic information is present in the features.

In our experience, this bias is most visually pronounced in regions with weak semantic content, for example background regions. In those cases, the semantic signal is weaker, and the positional component of the features can dominate the similarity map more clearly. For object regions such as the cat in your example, the semantic signal is much stronger, so the positional response may still be present but visually less pronounced.

Additionally, in segmentation, the reference descriptor is a prototype obtained by averaging features over the foreground object mask. If the object is reasonably large, as in your example, this averaging combines features from many spatial locations. As a result, part of the positional component is already averaged out before computing the target similarity map. Therefore, this visualization is not always the best setup to visually highlight the positional bias, even though removing it still gives consistent numerical gains.

This is why, in the paper, we also focus on semantic correspondence, where the effect is better isolated by using single patch features to compute similarity. For this reason, I would especially suggest looking at the [recently added semantic correspondence](https://github.com/visinf/INSID3#-semantic-correspondence) section in the repository. In that setting, the descriptor is sampled at the keypoint level rather than averaged over a full object mask, which makes the positional effect much easier to isolate (with improvements of up to +7.5 PCK)

It is also worth noting that, depending on the model size, this effect can be more or less evident from a qualitative point of view. However, this does not mean that the positional bias is absent or that it is not being removed. For example, with the `Large` model, the qualitative difference can be less visually striking, but debiasing still gives a gain of +2.2  (`Table 2` of the paper).

So, in short, the debiasing effect is not guaranteed to look dramatic in every individual similarity map. The effect is usually clearer when looking at single-patch matching, or at low-semantic-content background/texture regions. For segmentation, the benefit is better assessed statistically over datasets, where we observe consistent gains.

In general, the bias is systematic, but its visual magnitude can vary quite a lot depending on the model scale, the image pair, and the amount of semantic information encoded in the selected features.

Please let me know if you have any other doubts 😄

**Comment by Ginven-W** [comment]
Hi @ClaudiaCuttano ,thank you very much for your clear and detailed explanation! 
It fully answers my questions. I have no further doubts for now.
Really appreciate your help! 😊
